### PR TITLE
Merging `maybeOverlappingIndices` and `has_internal_overlap`

### PR DESCRIPTION
**Pull request by jjsjann123**
These two functions basically serves the same purpose.
`has_internal_overlap` provides a more flexible API (where
`MemOverlap::TOO_HARD` is provided), while `maybeOverlappingIndices` provides
meaningful flags on more general cases.

We get the best out of each and unifies the API.

Test cases are also provided. One thing to note here is, we should NEVER assert
on return being `MemOverlap::TOO_HARD`, which would block future improvement to
the `has_internal_overlap` function to cover more grounds.
We should instead assert on not returning the incorrect flags for cases not yet
supported by current implementation.

### CODE REVIEW

**Comment by SsnL** [comment]
could https://github.com/pytorch/pytorch/blob/3fe00f0c906ab4166058a8619b3c2edf121b2537/tools/autograd/templates/Functions.cpp#L1387
be also merged by any chance? I remember @zou3519 telling me that it does a different thing though...

**Comment by jjsjann123** [comment]
The implementation is slightly different, but I think it's doing the same thing. 
The implementation here didn't special case tensor with a dimension of size 0 here.

Just to be on the safe side, let me take a careful look at that `NOTE [ Detecting Memory Overlap Within A Strided Tensor ]` before merging it :)

File: aten/src/ATen/MemoryOverlap.cpp
Comment (VitalyFedyunin) [inline aten/src/ATen/MemoryOverlap.cpp]: Returns `MemOverlap::NO`

File: aten/src/ATen/MemoryOverlap.cpp
Comment (VitalyFedyunin) [inline aten/src/ATen/MemoryOverlap.cpp]: dims <= 1 tensor can't have overlaps, right?

File: aten/src/ATen/cuda/CUDAApplyUtils.cuh
Comment (VitalyFedyunin) [inline aten/src/ATen/cuda/CUDAApplyUtils.cuh]: returns `MemOverlap::NO`

File: aten/src/ATen/MemoryOverlap.cpp
Comment (zou3519) [reply aten/src/ATen/MemoryOverlap.cpp]: `x = torch.tensor(3.).expand(100)` has dim 1 and internal memory overlap

Comment (VitalyFedyunin) [reply aten/src/ATen/MemoryOverlap.cpp]: (facepalm)

File: aten/src/ATen/MemoryOverlap.cpp
Comment (apaszke) [inline aten/src/ATen/MemoryOverlap.cpp]: Ugh, can we use `std::sort` instead of `qsort`? That has a better chance of e.g. inlining the comparator.

**Comment by jjsjann123** [comment]
Realized the logic in the function i merged earlier was wrong. It should have be doing the same thing that @SsnL pointed out.

Let me merge all of them together.

**Comment by SsnL** [comment]
For the function I referenced earlier, it is kind of tricky because part of the algorithm is implemented in another function at 
https://github.com/pytorch/pytorch/blob/3fe00f0c906ab4166058a8619b3c2edf121b2537/tools/autograd/templates/Functions.cpp#L1428

The full algorithm is described at
https://github.com/pytorch/pytorch/blob/3fe00f0c906ab4166058a8619b3c2edf121b2537/tools/autograd/templates/Functions.cpp#L1190-L1201

The rest of that long NOTE is just proofs on how the property holds for all **current** strided tensor view ops, except for `as_strided` and `unfold`.

Since the implemented is split into two functions (for perf reasons), feel free to skip merging that if it becomes too hard. I can give it a try later. But if you are going to follow the same (or a similar algorithm), I hope you can move the note to the new place. Thank you. :)

**Comment by ezyang** [comment]
@VitalyFedyunin @SsnL  are there any further blockers on this?

**Comment by VitalyFedyunin** [comment]
@ezyang my comments not addressed

**Review by ezyang — CHANGES_REQUESTED** [review]
comments are not addressed

**Comment by jjsjann123** [comment]
Simon's request is not address neither. I'm currently occupied on other stuff right now, I'll get back to this at some point and put an update when it's ready for another review. Apology for the delay 😢

**Comment by jjsjann123** [comment]
Sorry for the delayed response. But address the request from @SsnL and typos that @VitalyFedyunin got.

**Comment by VitalyFedyunin** [comment]
Hi! Can you please rebase and attach Before/After benchmarks, thanks.

**Comment by jjsjann123** [comment]
Two things this PR does:

1. The main idea is to fix: https://github.com/pytorch/pytorch/blob/949678bd9e82c12efe77fb0fb5471a17ded128f4/aten/src/ATen/cuda/detail/IndexUtils.cu#L64-L68
`maybeOverlappingIndices` here is basically wrong. It would return `false` with tensor (size: [2, 2, 10], stride: [11, 10, 1]), while the overlapping could be detected using `has_internal_overlap`;

2. Improve the check for `has_internal_overlap`: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/MemoryOverlap.cpp#L10-L24
Pushing more cases from `TOO_HARD` to `NO` and `YES` repectively.
e.g.
```
x = torch.randn(4)

# output has been updated from `TOO_HARD` to `NO`
print(torch._debug_has_internal_overlap(x[::2]))

# output has been updated from `TOO_HARD` to `YES`
z = x.as_strided((2, 2), (1, 1))
```

**Comment by jjsjann123** [comment]
So something naive as:
```
  1 import torch
  2 import timeit
  3
  4 x = torch.randn(128, 64, 128, 128)
  5
  6 setup = """
  7 import torch
  8 x = torch.randn(128, 64, 128, 128)
  9 """
 10 stmt = """
 11 torch._debug_has_internal_overlap(x[::2, ::2, ::2, ::2])
 12 torch._debug_has_internal_overlap(x[:, ::2, ::2, ::2])
 13 torch._debug_has_internal_overlap(x.as_strided((2,2,2),(1,1,1)))
 14 """
 15
 16 print(timeit.timeit(stmt=stmt, setup=setup, number=10000))
```
gives ~0.140 on my current branch.
Same code gives ~0.122 on docker hub 1.3-runtime (14 days old `c3f6590d7263`) & ~0.142 on docker hub 1.3-cuda10.1-cudnn7-devel (14 days old `d22bdcc867e2`)

I'll be surprised that simple host code like this would have impact on performance.

Meanwhile, it did improve our error checking on memory overlap.
```
import torch
x = torch.zeros(3)
y = x.as_strided((2, 2), (1, 1))

# It now throws an error for multiple element being written-to.
# which is desired behavior for tensor iterator
y += 1
```

**Comment by jjsjann123** [comment]
Failure from compatibility test:

```Oct 29 03:17:23 processing existing schema:  aten::full_like(Tensor self, Scalar fill_value, *, int? memory_format=None) -> (Tensor)
Oct 29 03:17:23 Can NOT find backward compatible schemas after changes for schema aten::full_like(Tensor self, Scalar fill_value, *, int? memory_format=None) -> (Tensor) from the following candidates:
Oct 29 03:17:23 [
Oct 29 03:17:23 aten::full_like(Tensor self, Scalar fill_value) -> (Tensor)
Oct 29 03:17:23 	aten::full_like.dtype(Tensor self, Scalar fill_value, *, int dtype, int layout, Device device, bool pin_memory=False) -> (Tensor)
Oct 29 03:17:23 ]
Oct 29 03:17:23 The PR is introducing backward incompatible changes to the operator library. Please contact PyTorch team to confirm whether this change is wanted or not.```

Looks like something related to previous NHWC changes and not related to my change.

**Comment by jjsjann123** [comment]
@VitalyFedyunin Requests have been addressed. Please take another look :bowing_man:

**Comment by jjsjann123** [comment]
I can try to resolve the conflicts again. Is there any remaining issue that is blocking the merge? I'd like to resolve that when merging as well.

**Comment by VitalyFedyunin** [comment]
Please hold on merge, I will do my best to provide you with detailed review by Monday (Nov 18).

**Pull request closed without merging**
